doc/experimental_ui_api.md

```diff
@@ -0,0 +1,2027 @@
+with the same window type. Additionally, all persistent properties will be
+changed by this operation, just like `open()`.
+
+### State and parameter tables
+
+Since UIs generally have some state associated with each open window, contexts
+provide a means of holding state across calls to the builder function via a
+_state table_. When a `ui.Context` is constructed, a state table can be
+provided that holds state that will be passed to the builder function every
+time it is called for this UI. This table can be modified by the builder
+function or by event handlers in the UI. The UI API will never modify a state
+table by itself. The state table can be obtained from a context via the
+`get_state()` method or replaced entirely via `set_state()`.
+
+The following is a basic example that uses a state table:
+
+```lua
+local function builder(context, player, state, param)
+    return ui.Window "gui" {
+        root = ui.Root {
+            size = {108, 40},
+            padding = {4},
+
+            box_fill = "black#8C",
+
+            ui.Button "counter" {
+                box_fill = "maroon",
+                label = "Clicks: " .. state.num_clicks,
+
+                on_press = function(ev)
+                    state.num_clicks = state.num_clicks + 1
+                    context:update()
+                end,
+            },
+        },
+    }
+end
+
+core.register_on_joinplayer(function(player)
+    local state = {num_clicks = 0}
+    ui.Context(builder, player:get_player_name(), state):open()
+end)
+```
+
+The state table is primarily for persistent data. However, it is often useful
+to send temporary data that only applies to a single call of the builder
+function, such as when setting persistent fields. This can be provided in the
+form of a _parameter table_. Parameter tables can be provided to the methods
+`open()`, `update()`, and `reopen()` and will be passed directly to the builder
+function. After the builder function returns, the parameter table is discarded.
+
+The following is a basic example that sets a persistent field when first
+opening the window by using parameter tables (note that it uses a slider, which
+is not implemented yet):
+
+```lua
+local function builder(context, player, state, param)
+    return ui.Window "gui" {
+        root = ui.Root {
+            size = {500, 40},
+            padding = {4},
+
+            box_fill = "black#8C",
+
+            ui.Slider "percent" {
+                label = state.scroll .. "%",
+
+                min = 0,
+                max = 100,
+
+                -- `param.set_scroll` is only set in `open()`, not in the
+                -- `update()` in `on_scroll`, so `value` is only set to
+                -- `state.scroll` when the window is first opened.
+                value = param.set_scroll and state.scroll,
+
+                on_scroll = function(ev)
+                    state.scroll = ev.value
+                    context:update()
+                end,
```

Comment:
So, scrolling is server side? or will it be both?

I'm thinking that both would allow lazy loading, I guess, can't think of many other use cases yet.

Reply:
Scrolling will be client-side; I'm just storing the scroll value on the server for the sole purpose of displaying a label.

Reply:
> or will it be both?

but the server does have an on_scroll callback, then?

Reply:
It will when scrollbars exist :) They don't yet, as mentioned in the paragraph above the example.